### DISCUSSION TITLE
clipping and noising

### DISCUSSION THREAD

**Discussion by npapernot** [NVIDIA/NVFlare#484] [Q&A]
Hello, 

I'm curious about the following line:

https://github.com/NVIDIA/NVFlare/blob/0753aba7ad97e09db474e0d56ce843e4eee8285a/nvflare/app_common/filters/svt_privacy.py#L119

Could you explain why the noising is done within the clipping operation? How do you analyze the sensitivity of the algorithm given that?

Thank you!

**Comment by wyli** [comment]
The clipping was not done in the original SVT paper https://arxiv.org/abs/1603.01699, but it's used in the tumour segmentation paper https://arxiv.org/abs/1910.00962 "Gradient clipping is also applied, which is a widely-used method, acting as a model regulariser to prevent over-fitting.". In my understanding computing the sensitivity using gamma in this case becomes a more conservative (but bounded) estimation.

(converting this into a discussion)

**Reply by msheller** [reply]
@wyli @holgerroth I don't think this answers the question from @npapernot 

If the sensitivity calculation is relying on the clip, the algorithm needs to add noise after the clip, e.g.:

delta_w[accepted] = np.clip(delta_w[accepted] + noise, a_min=-self.gamma, a_max=self.gamma)
delta_w[accepted] = delta_w[accepted] + noise

(Note: in this case, I would remove the "+noise" inside the clip, since it would serve no purpose I can think of).

If the sensitivity calculation is not relying on the clip, how is it calculated? (This was @npapernot 's original question, as I read it).

@npapernot , does this match your thinking?

Note that in the SVT paper the sensitivity is defined as a precondition, so they have already bound the queries in Q and you won't see a clip on the query inside the algorithm. Even so, they state that outputting q_i+v_i is non-private and you must emit either "above" or "below" threshold. I would need to re-read to recall why on this second point.

**Reply by npapernot** [reply]
Indeed, this does not answer my initial question as I was wondering the following: if the sensitivity calculation is not relying on the clip, how is it calculated?

**Reply by wyli** [reply]
in my understanding, the sensitivity defined in this context is that given two training datasets that are neighbours (diff by one training item), the released model difference delta_w will differ at most gamma because of the clipping operation, so I think gamma is a valid upper bound for sensitivity. @npapernot would be great to have your further insight on this.

**Reply by npapernot** [reply]
But gamma is the sensitivity after clipping, not before, so I'm not sure I understand how the noise is calibrated according to the sensitivity.

**Reply by wyli** [reply]
Yes while it's helpful for stabilising delta_w for model training, I agree the current implementation makes privacy budget quantification less straightforward. cc @holgerroth Perhaps we could make this optional, providing the user choices of clipping(delta_w) + noise? (also we could have the options of clipping by value or clipping by norm. although in my experience, the model convergence can become heavily dependent on the choice of these parameters.)

**Reply by msheller** [reply]
The original SVT paper requires the input stream Q have sensitivity delta as a precondition (see the caption under Fig. 1). Answering @npapernot 's question is required by the algorithm.
